prow/bugzilla/client.go

```diff
@@ -214,37 +231,112 @@ func getRootForClone(c Client, bug *Bug) (*Bug, error) {
 
 // GetRootForClone returns the original bug.
 func (c *client) GetRootForClone(bug *Bug) (*Bug, error) {
-	return getRootForClone(c, bug)
+	bugCache := map[int]*Bug{}
+	return getRootForClone(c, bug, bugCache)
 }
 
 // GetAllClones returns all the clones of the bug including itself
 // Differs from GetClones as GetClones only gets the child clones which are one level lower
-func (c *client) GetAllClones(bug *Bug) ([]*Bug, error) {
-	return getAllClones(c, bug)
+func (c *client) GetAllClones(bugID int) ([]*Bug, *Bug, *Bug, error) {
+	bugCache := map[int]*Bug{}
+	return getAllClones(c, bugID, bugCache)
 }
-func getAllClones(c Client, bug *Bug) ([]*Bug, error) {
-	root, err := getRootForClone(c, bug)
+
+type clonesCacheWithMux struct {
+	cache map[int][]int
+	mux   sync.Mutex
+}
+
+var clonesCache = clonesCacheWithMux{cache: map[int][]int{}}
+
+func getAllClones(c Client, bugID int, intraCallCache map[int]*Bug) ([]*Bug, *Bug, *Bug, error) {
+
+	// Global clones cache should only be accessed after ensuring proper thread safety
+	clonesCache.mux.Lock()
+	clonesID, ok := clonesCache.cache[bugID]
+	clonesCache.mux.Unlock()
+	var bug *Bug
+	var getBugErr error
+	var wg sync.WaitGroup
+	cloneErrors := make([]error, len(clonesID))
+	clones := make([]*Bug, len(clonesID))
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// This error is being handled after the wg.Wait
+		bug, getBugErr = c.GetBug(bugID)
+	}()
+	if ok {
+		var intraCallCacheMux sync.Mutex
```

Comment:
its a mutex btw, a mux is something else (using used as a term to refer to a http router)